File: src/lambdas/calimaco/get-user-detail/get-user-detail.py
import json
import logging
import requests
import boto3
import os
from typing import Dict, Any

FN_CRYPTO = os.environ.get('FN_CRYPTO')

# Configuración del logger
logger = logging.getLogger()
logger.setLevel(logging.INFO)

lambda_client = boto3.client('lambda')

# Lambda Invoke
def invoke_crypto(session: str) -> str:
    """
    Invoca la función Lambda Crypto.
    """
    try:
        body = {
            'mode': "decrypt",
            'payload': session
        }

        logger.info("Datos enviados a crypto-function %s", json.dumps(body, indent=4))

        response = lambda_client.invoke(
            FunctionName=FN_CRYPTO,
            InvocationType='RequestResponse',       
            Payload=json.dumps(body)
        )

        payload = json.loads(response['Payload'].read())
        logger.info("Respuesta de crypto_function: %s", json.dumps(payload, indent=4))

        return payload.get('body', "")
    except Exception as ex:
        logger.error("Error al invocar crypto-function: %s", str(ex))

def decode_jwt_from_event(event):
    try:
        token = event["requestContext"]["authorizer"]["claims"]
        return token
    except KeyError:
        return {"error": "No se pudo obtener el token del contexto"}

def lambda_handler(event: Dict[str, Any], context: Any):

    try:

        token_payload = decode_jwt_from_event(event)
        if "error" in token_payload:
            return {
                'statusCode': 401,
                'headers': {
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Headers": "*",
                    "Access-Control-Allow-Methods": "OPTIONS,GET",
                    "Content-Type": "application/json"
                },
                'body': json.dumps({"message": "Sin Autorización"})
            }

        session = token_payload.get("session_col", "")
        logger.info(f"session {session}")

        sessionCol = invoke_crypto(session)
        logger.info(sessionCol)
        
        url = "https://api.casinoatlanticcity.com/api/data/getUserDetails"

        payload = f"company=ACP&session={sessionCol}"
        headers = {
            'accept': 'application/json, text/plain, /',
            'access-control-allow-origin': '*',
            'content-type': 'application/x-www-form-urlencoded; charset=UTF-8',
        }
        logger.info("Datos enviados a USER-DETAILS %s", json.dumps(payload, indent=4))
        response = requests.request("POST", url, headers=headers, data=payload)

        user_detail = response.json()
        logger.info(f"response.json {user_detail}")

        if user_detail.get("code", 0) != 0:
            return {
                'statusCode': 401,
                'headers': {
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Headers": "*",
                    "Access-Control-Allow-Methods": "OPTIONS,GET",
                    "Content-Type": "application/json"
                },
                'body': json.dumps({"message": "Sin Autorización"})
            }

        try:
            accounts_list = user_detail.get("user", {}).get("accounts", [])
            
            if accounts_list:
                for account in accounts_list:
                    original_amount = account.get('amount')
                    
                    if isinstance(original_amount, (int, float)):
                        new_amount_float = original_amount / 100.0
                        
                        account['amount'] = round(new_amount_float, 2)
                        
                    elif original_amount is None:
                        account['amount'] = 0.0

        except Exception as parse_ex:
            logger.error(f"Error al parsear decimales de 'accounts': {str(parse_ex)}")

        return {
                'statusCode': 200,
                'headers': {
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Headers": "*",
                    "Access-Control-Allow-Methods": "OPTIONS,GET",
                    "Content-Type": "application/json"
                },
                'body': json.dumps(user_detail)
            }

    except Exception as e:
        logger.error(f"Error en la solicitud: {e}")
        return {
            'statusCode': 500,
            'headers': {
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "*",
                "Access-Control-Allow-Methods": "OPTIONS,GET",
                "Content-Type": "application/json"
            },
            'body': json.dumps({
                'message': 'Ocurrió un error inesperado'
            })
        }

File: src/lambdas/calimaco/get-user-detail.mjs
import { LambdaClient, InvokeCommand } from "@aws-sdk/client-lambda";

const FN_CRYPTO = process.env.FN_CRYPTO;
const REGION = process.env.REGION;
const CALIMACO_BASE_URL = process.env.CALIMACO_BASE_URL;

const lambdaClient = new LambdaClient({ region: REGION });

async function invokeCrypto(session) {
    try {
        const command = new InvokeCommand({
            FunctionName: FN_CRYPTO,
            InvocationType: "RequestResponse",
            Payload: JSON.stringify({ mode: "decrypt", payload: session })
        });

        const response = await lambdaClient.send(command);
        const result = JSON.parse(new TextDecoder().decode(response.Payload));
        return result.body || "";
    } catch (error) {
        console.error("Error invoking crypto lambda:", error);
        return "";
    }
}

function decodeJwtFromEvent(event) {
    try {
        return event.requestContext.authorizer.claims;
    } catch (error) {
        return { error: "No se pudo obtener el token del contexto" };
    }
}

function createResponse(statusCode, data) {
    return {
        statusCode,
        headers: {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Headers": "*",
            "Access-Control-Allow-Methods": "OPTIONS,GET",
            "Content-Type": "application/json"
        },
        body: JSON.stringify(data)
    };
}

export const lambda_handler = async (event) => {
    try {
        const tokenPayload = decodeJwtFromEvent(event);
        if (tokenPayload.error) {
            return createResponse(401, { message: "Sin Autorización" });
        }

        const session = tokenPayload.session_col || "";
        console.log("session:", session);

        const sessionCol = await invokeCrypto(session);
        console.log("sessionCol:", sessionCol);

        const response = await fetch(`${CALIMACO_BASE_URL}/data/getUserDetails`, {
            method: "POST",
            headers: {
                "accept": "application/json, text/plain, */*",
                "access-control-allow-origin": "*",
                "content-type": "application/x-www-form-urlencoded; charset=UTF-8"
            },
            body: new URLSearchParams({
                company: "ACP",
                session: sessionCol
            })
        });

        const userDetail = await response.json();
        console.log("response.json:", userDetail);

        if (userDetail.code && userDetail.code !== 0) {
            return createResponse(401, { message: "Sin Autorización" });
        }

        // Parse accounts amounts (divide by 100)
        try {
            const accountsList = userDetail.user?.accounts || [];
            
            if (accountsList.length > 0) {
                for (const account of accountsList) {
                    const originalAmount = account.amount;
                    
                    if (typeof originalAmount === "number") {
                        account.amount = Math.round((originalAmount / 100) * 100) / 100; // Round to 2 decimals
                    } else if (originalAmount === null || originalAmount === undefined) {
                        account.amount = 0.0;
                    }
                }
            }
        } catch (parseError) {
            console.error("Error parsing accounts decimals:", parseError);
        }

        return createResponse(200, userDetail);

    } catch (error) {
        console.error("Error en la solicitud:", error);
        return createResponse(500, { message: "Ocurrió un error inesperado" });
    }
};
